extensions/vi-mode/options.lisp

```diff
@@ -0,0 +1,178 @@
+(defpackage #:lem-vi-mode/options
+  (:use #:cl)
+  (:import-from #:lem-vi-mode/core
+                #:change-directory)
+  (:import-from #:parse-number
+                #:parse-number)
+  (:import-from #:cl-ppcre
+                #:scan-to-strings)
+  (:import-from #:alexandria
+                #:if-let
+                #:when-let
+                #:once-only
+                #:with-gensyms)
+  (:export #:define-vi-option
+           #:get-option
+           #:vi-option
+           #:vi-option-name
+           #:vi-option-value
+           #:vi-option-default
+           #:vi-option-type
+           #:vi-option-aliases
+           #:vi-option-getter
+           #:vi-option-set-hook
+           #:vi-option-documentation
+           #:reset-vi-option-value
+           #:toggle-vi-option-value
+           #:execute-set-command
+
+           ;; Options
+           #:autochdir))
+(in-package #:lem-vi-mode/options)
+
+(defstruct vi-option
+  (name nil :type string)
+  %value
+  default
+  (type t :type (member t boolean number string))
+  (aliases '() :type list)
+  (getter nil :type (or function null))
+  (set-hook nil :type (or function null))
+  (documentation nil :type (or string null)))
+
+(defvar *options* (make-hash-table :test 'equal))
+(defvar *option-aliases* (make-hash-table :test 'equal))
+
+(defun canonical-option-name (name)
+  (or (gethash name *option-aliases*)
+      name))
+
+(defun get-option (name &optional (error-if-not-exists t))
+  (check-type name string)
+  (let ((name (canonical-option-name name)))
+    (multiple-value-bind (option exists)
+        (gethash name *options*)
+      (when (and (null exists)
+                 error-if-not-exists)
+        (error "Unknown option: ~A" name))
```

Comment:
error is handled as an internal error.
Instead of treating user input as an internal error, it may be better to use editor-error or a condition that inherits from it.